aws-lc-rs/src/key_wrap.rs

```diff
@@ -0,0 +1,354 @@
+                // * max_out >= in_len - 8
+                if 1 != indicator_check!(unsafe {
+                    AES_unwrap_key_padded(
+                        &aes_key,
+                        output.as_mut_ptr(),
+                        &mut out_len,
+                        output.len(),
+                        input.as_ptr(),
+                        input.len(),
+                    )
+                }) {
+                    return Err(Unspecified);
+                };
+
+                out_len
+            }
+            WrappingMode::Unpadded => {
+                // AWS-LC validates the following:
+                // * in_len < INT_MAX
+                // * in_len > 24
+                // * in_len % 8 == 0
+                let out_len = indicator_check!(unsafe {
+                    AES_unwrap_key(
+                        &aes_key,
+                        null(),
+                        output.as_mut_ptr(),
+                        input.as_ptr(),
+                        input.len(),
+                    )
+                });
+
+                if out_len == -1 {
+                    return Err(Unspecified);
+                }
+
+                let out_len: usize = out_len.try_into().map_err(|_| Unspecified)?;
+
+                debug_assert_eq!(out_len, input.len() - 8);
+
+                out_len
+            }
+        };
+
+        Ok(&mut output[..out_len])
+    }
+}
+
+/// The key-encryption key used with the selected cipher algorithn to wrap or unwrap a key.
+pub struct KeyEncryptionKey {
+    algorithm: &'static Algorithm,
+    key: Box<[u8]>,
+    mode: WrappingMode,
+}
+
+impl KeyEncryptionKey {
+    /// Creates a new `KeyEncryptionKey` using the provider cipher algorithm, key bytes, and wrapping mode.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if `key_bytes.len()` does not match the size expected for the provided algorithm.
+    pub fn new(
+        algorithm: &'static Algorithm,
+        key_bytes: &[u8],
+        mode: WrappingMode,
+    ) -> Result<Self, Unspecified> {
+        if algorithm.key_len() != key_bytes.len() {
+            return Err(Unspecified);
+        }
+
+        let key = Vec::from(key_bytes).into_boxed_slice();
+
+        Ok(Self {
+            algorithm,
+            key,
+            mode,
+        })
+    }
+
+    /// Peforms the key wrap encryption algorithm using `KeyEncryptionKey`'s configured cipher algorithm
+    /// and wrapping operation mode. It wraps the provided `input` plaintext and writes the
+    /// ciphertext to `output`.
+    ///
+    /// If `WrappingMode::Unpadded` is the configured mode, then `input.len()` must be a multiple of 8 and non-zero.
+    ///
+    /// # Sizing `output`
+    /// `output` must be sized appropriately depending on the configured [`WrappingMode`].
+    /// * [`WrappingMode::Padded`]: `output.len() >= (input.len() + 15)`
+    /// * [`WrappingMode::Unpadded`]: `output.len() >= (input.len() + 8)`
```

Comment:
With the input and output pre-/post- conditions differing between the modes, I do wonder whether it's a mistake to combine them behind the same interface.